x-pack/filebeat/input/httpjson/httpjson_test.go

```diff
@@ -6,783 +6,409 @@ package httpjson
+				"date_cursor.initial_interval": "10m",
+				"date_cursor.date_format":      "2006-01-02T15:04:05Z",
+			},
+			handler: dateCursorHandler(),
+			expected: []string{
+				`{"@timestamp":"2002-10-02T15:00:00Z","foo":"bar"}`,
+				`{"@timestamp":"2002-10-02T15:00:01Z","foo":"bar"}`,
+			},
+			duration: 700 * time.Millisecond,
+			setup: func(map[string]interface{}) interface{} {
+				timeNow = func() time.Time {
+					t, _ := time.Parse(time.RFC3339, "2002-10-02T15:00:00Z")
+					return t
+				}
+				return nil
+			},
+		},
+		{
+			name: "Test pagination",
+			baseConfig: map[string]interface{}{
+				"http_method":          "GET",
+				"interval":             0,
+				"pagination.id_field":  "nextPageToken",
+				"pagination.url_field": "page",
+				"json_objects_array":   "items",
+			},
+			handler:  paginationHandler(),
+			expected: []string{`{"foo":"bar"}`, `{"foo":"bar"}`},
+		},
+		{
+			name: "Test oauth2",
+			baseConfig: map[string]interface{}{
+				"http_method":          "POST",
+				"interval":             "0",
+				"oauth2.client.id":     "a_client_id",
+				"oauth2.client.secret": "a_client_secret",
+				"oauth2.endpoint_params": map[string]interface{}{
+					"param1": "v1",
 				},
-				{"bar": "foo"},
+				"oauth2.scopes": []string{"scope1", "scope2"},
 			},
-		}
-		b, _ := json.Marshal(message)
-		w.WriteHeader(http.StatusOK)
-		w.Write(b)
-	}))
-}
-
-func runTest(t *testing.T, ts *httptest.Server, m map[string]interface{}, run func(input *HttpjsonInput, out *stubOutleter, t *testing.T)) {
-	testSetup(t)
-	defer ts.Close()
-	m["url"] = ts.URL
-	cfg := common.MustNewConfigFrom(m)
-	// Simulate input.Context from Filebeat input runner.
-	inputCtx := newInputContext()
-	defer close(inputCtx.Done)
-
-	// Stub outlet for receiving events generated by the input.
-	eventOutlet := newStubOutlet()
-	defer eventOutlet.Close()
-
-	connector := channel.ConnectorFunc(func(_ *common.Config, _ beat.ClientConfig) (channel.Outleter, error) {
-		return eventOutlet, nil
-	})
-
-	in, err := NewInput(cfg, connector, inputCtx)
-	if err != nil {
-		t.Fatal(err)
-	}
-	input := in.(*HttpjsonInput)
-	defer input.Stop()
-
-	run(input, eventOutlet, t)
-}
-
-func newInputContext() input.Context {
-	return input.Context{
-		Done: make(chan struct{}),
-	}
-}
-
-type stubOutleter struct {
-	sync.Mutex
-	cond   *sync.Cond
-	done   bool
-	Events []beat.Event
-}
-
-func newStubOutlet() *stubOutleter {
-	o := &stubOutleter{}
-	o.cond = sync.NewCond(o)
-	return o
-}
-
-func (o *stubOutleter) waitForEvents(numEvents int) ([]beat.Event, bool) {
-	o.Lock()
-	defer o.Unlock()
-
-	for len(o.Events) < numEvents && !o.done {
-		o.cond.Wait()
+			setup: func(config map[string]interface{}) interface{} {
+				server := httptest.NewServer(http.HandlerFunc(oauth2TokenHandler))
+				config["oauth2.token_url"] = server.URL
+				return server
+			},
+			teardown: func(i interface{}) {
+				server := i.(*httptest.Server)
+				server.Close()
+			},
```

Comment:
do we ever return some other type but an http server?

teardown as is can be removed. Better pass `testing.T` to setup. the `close` can be registered with the test via `t.Cleanup`.

For this tests we end up with 2 active servers. how about only providing a `setupServer` method? then `ssl`, `setup`, and `teardown` can be replaced by a single replacable function.